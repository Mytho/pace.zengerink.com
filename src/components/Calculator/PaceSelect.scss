@import './Calculator.scss';

.pace-select {
    @include calculator-row;
}