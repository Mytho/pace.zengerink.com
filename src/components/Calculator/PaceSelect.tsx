import React from 'react'
import NumberSelect from './NumberSelect'

export enum PaceFragment {
  Minutes = 'minutes',
  Seconds = 'seconds',
}

interface PaceSelectProps {
  value: number
  onChange: (value: number) => void
}

interface Pace {
  [key: string]: number
}

export default class DurationSelect extends React.Component<PaceSelectProps, {}> {
  handleChange(prop: string, value: number) {
    const pace = this.toPace(this.props.value)
    pace[prop] = value
    this.props.onChange(this.toSeconds(pace))
  }

  toSeconds(pace: Pace): number {
    return pace[PaceFragment.Minutes] * 60 + pace[PaceFragment.Seconds]
  }

  toPace(seconds: number): Pace {
    return {
      [PaceFragment.Minutes]: Math.floor(seconds / 60),
      [PaceFragment.Seconds]: seconds % 60,
    }
  }

  render() {
    const pace = this.toPace(this.props.value)

    return (
      <div className="duration-select">
        <NumberSelect
          name={PaceFragment.Minutes}
          value={pace[PaceFragment.Minutes]}
          max="59"
          onChange={this.handleChange.bind(this, PaceFragment.Minutes)}
        />
        <span>:</span>
        <NumberSelect
          name={PaceFragment.Seconds}
          value={pace[PaceFragment.Seconds]}
          max="59"
          pad="2"
          onChange={this.handleChange.bind(this, PaceFragment.Seconds)}
        />
        <span>min/km</span>
      </div>
    )
  }
}
