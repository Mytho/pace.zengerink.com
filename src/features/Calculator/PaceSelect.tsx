import React, { ReactElement } from 'react'
import NumberSelect from './NumberSelect'
import './PaceSelect.scss'

type PaceFragmentName = 'minutes' | 'seconds'

interface PaceSelectProps {
  label: string
  value: number
  onChange: (value: number) => void
}

interface PaceFragments {
  minutes: number
  seconds: number
}

const toSeconds = (pace: PaceFragments): number => pace.minutes * 60 + pace.seconds
const toPace = (seconds: number): PaceFragments => ({ minutes: Math.floor(seconds / 60), seconds: seconds % 60 })

const PaceSelect = (props: PaceSelectProps): ReactElement => {
  const handleChange = (fragmentName: PaceFragmentName, value: number) => {
    const fragments = toPace(props.value)
    fragments[fragmentName] = value
    props.onChange(toSeconds(fragments))
  }

  const paceFragments = toPace(props.value)

  return (
    <div className="pace-select">
      <div className="pace-select__label">{props.label}</div>
      <div className="pace-select__input">
        <NumberSelect
          name="minutes"
          value={paceFragments.minutes}
          max="59"
          onChange={(value) => handleChange('minutes', value)}
        />
        <span className="pace-select__separator">:</span>
        <NumberSelect
          name="seconds"
          value={paceFragments.seconds}
          max="59"
          pad="2"
          onChange={(value) => handleChange('seconds', value)}
        />
        <span className="pace-select__value-label">min/km</span>
      </div>
    </div>
  )
}

export default PaceSelect
