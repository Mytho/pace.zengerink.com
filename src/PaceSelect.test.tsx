import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import PaceSelect, { PaceFragment } from './PaceSelect'


describe('render', () => {
  it('renders correctly', () => {
    const onChange = jest.fn()
    const wrapper = shallow(<PaceSelect value={3661} onChange={onChange} />)
    const json = shallowToJson(wrapper)
    expect(json).toMatchSnapshot()
  })
})

describe('onChange', () => {
  const testOnChange = (name: string, newValue: number, expectedArgument: number) => {
    const onChange = jest.fn()
    const wrapper = shallow(<PaceSelect value={0} onChange={onChange} />)
    const select = wrapper.find(`[name="${name}"]`)
    expect(select.length).toBe(1)

    select.simulate('change', newValue)
    expect(onChange.mock.calls.length).toBe(1)
    expect(onChange.mock.calls[0][0]).toBe(expectedArgument)
  }

  it('is called when one of the selects changes', () => {
    testOnChange(PaceFragment.Minutes, 1, 60)
    testOnChange(PaceFragment.Seconds, 2, 2)
  })
})