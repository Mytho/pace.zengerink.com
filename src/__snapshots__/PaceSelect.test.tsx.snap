// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders correctly 1`] = `
<div
  className="duration-select"
>
  <NumberSelect
    max="59"
    name="minutes"
    onChange={[Function]}
    value={61}
  />
  <span>
    :
  </span>
  <NumberSelect
    max="59"
    name="seconds"
    onChange={[Function]}
    pad="2"
    value={1}
  />
  <span>
    min/km
  </span>
</div>
`;
